packages/babel-parser/src/parser/expression.js

```diff
@@ -2514,65 +2578,55 @@ export default class ExpressionParser extends LValParser {
     return this.finishNode(node, "YieldExpression");
   }
 
-  // Validates a pipeline (for any of the pipeline Babylon plugins) at the point
-  // of the infix operator `|>`.
-
-  checkPipelineAtInfixOperator(left: N.Expression, leftStartPos: number) {
-    if (this.getPluginOption("pipelineOperator", "proposal") === "smart") {
-      if (left.type === "SequenceExpression") {
-        // Ensure that the pipeline head is not a comma-delimited
-        // sequence expression.
-        this.raise(leftStartPos, Errors.PipelineHeadSequenceExpression);
-      }
-    }
-  }
-
-  parseSmartPipelineBody(
-    childExpression: N.Expression,
-    startPos: number,
-    startLoc: Position,
-  ): N.PipelineBody {
-    this.checkSmartPipelineBodyEarlyErrors(childExpression, startPos);
-
-    return this.parseSmartPipelineBodyInStyle(
-      childExpression,
-      startPos,
-      startLoc,
-    );
-  }
-
-  checkSmartPipelineBodyEarlyErrors(
-    childExpression: N.Expression,
-    startPos: number,
-  ): void {
+  // This helper method is to be called immediately
+  // after a Hack-style pipe body is parsed.
+  // The `startPos` is the starting position of the pipe body.
+
+  checkHackPipeBodyEarlyErrors(startPos: number): void {
+    // If the following token is invalidly `=>`, `=`, etc.,
+    // then throw a human-friendly error
+    // instead of something like 'Unexpected token, expected ";"'.
+    // For example, `x => x |> y => #` (assuming `#` is the topic reference)
+    // groups into `x => (x |> y) => #`,
+    // and `(x |> y) => #` is an invalid arrow function.
+    // This is because Hack-style `|>` has tighter precedence than `=>`.
+    // (Unparenthesized `yield` expressions are handled
+    // in `parseHackPipeBody`.
+    // This is because `yield`’s “not allowed
+    // as identifier in generator” error will immediately
+    // occur before the pipe body is fully parsed.)
     if (this.match(tt.arrow)) {
-      // If the following token is invalidly `=>`, then throw a human-friendly error
-      // instead of something like 'Unexpected token, expected ";"'.
-      throw this.raise(this.state.start, Errors.PipelineBodyNoArrow);
-    } else if (childExpression.type === "SequenceExpression") {
-      this.raise(startPos, Errors.PipelineBodySequenceExpression);
+      throw this.raise(
+        this.state.start,
+        Errors.PipeBodyIsTight,
+        tt.arrow.label,
+      );
+    } else if (this.match(tt.assign)) {
+      throw this.raise(
+        this.state.start,
+        Errors.PipeBodyIsTight,
+        tt.arrow.assign,
```

Comment:
Should this be this?
```suggestion
        this.state.value,
```

Or maybe we could even unify both the arrow and assign cases:
```js
if (this.match(tt.arrow) || this.match(tt.assign)) {
  throw this.raise(this.state.start, Errors.PipeBodyIsTight, this.state.value);
}
```

Reply:
Unfortunately, I cannot figure out how to detect when an assignment is erroneously used without parentheses (for example, `a |> x = %`, which groups into `(a |> x) = %`), since the parser seems to eagerly consume the `=` before I can detect it at this point. So I’m just getting rid of trying to detect `tt.assign`.

Reply:
Maybe we could add the check in `parseMaybeAssign` (in the same place where we throw about the invalid LHS) but throw the better error message?

Or maybe better, we can check if `this.match(tt.assign)` after https://github.com/js-choi/babel/blob/c9823c517728a383b45387ab1b7205217fe67956/packages/babel-parser/src/parser/expression.js#L531.

Reply:
`parseMaybeAssign` turned out to be the best place to handle this, especially given the plurality of assignment operators. This has been resolved in cd8a457f77b75618f43dcb260f677e3b52620641.